src/coroutine/cancel.rs

```diff
@@ -32,27 +34,25 @@ impl CancelHandle {
 
     /// Poll to retrieve the exception thrown in the associated coroutine.
     pub fn poll_cancelled(&mut self, cx: &mut Context<'_>) -> Poll<PyObject> {
-        // SAFETY: only valid owned pointer are set in `ThrowCallback::throw`
-        let take = || unsafe {
+        let mut guard = self.0.waker.lock();
+        if self.is_cancelled() {
             // pointer cannot be null because it is checked the line before,
             // and the swap is protected by `&mut self`
-            Py::from_non_null(
-                NonNull::new(self.0.exception.swap(ptr::null_mut(), Ordering::Relaxed)).unwrap(),
-            )
-        };
-        if self.is_cancelled() {
-            return Poll::Ready(take());
+            let ptr = self.0.exception.swap(ptr::null_mut(), Ordering::Relaxed);
+            // pointer cannot be null because it is checked the line before,
+            // and the swap is protected by `&mut self`
+            // SAFETY: only valid owned pointer are set in `ThrowCallback::throw`
+            return Poll::Ready(unsafe { Py::from_non_null(NonNull::new(ptr).unwrap()) });
         }
-        self.0.waker.register(cx.waker());
-        if self.is_cancelled() {
-            return Poll::Ready(take());
+        if !matches!(guard.deref(), Some(waker) if cx.waker().will_wake(waker)) {
```

Comment:
I think this is too clever, i.e. checking whether the guard contains `Some` and whether the waker will be woken should be two nested if expressions.

Reply:
Do you want something like this?
```rust
if let Some(waker) = guard.deref() {
    if cx.waker().will_wake(waker) {
        return Poll::Pending;
    }
}
*guard = Some(cx.waker().clone());
Poll::Pending
```

Reply:
Yes, exactly this. I understand it is more text, but parsing the guard expression inside a negated `matches!` clauses is just making this unnecessarily difficult to read IMHO.

One other alternative I see would be something like

```rust
if guard.deref().map_or(false, |waker| cx.waker().will_wake(waker)) {
  return Poll::Pending;
}
```

but even this seems less readable than the two-clause variant.

Reply:
done